search/materialize.go

```diff
@@ -0,0 +1,404 @@
+	rg := m.cf.RowGroups()[rgi]
+	totalRows := int64(0)
+	for _, r := range rr {
+		totalRows += r.count
+	}
+	r := make([][]chunks.Meta, totalRows)
+
+	for i := minDataCol; i <= maxDataCol; i++ {
+		values, err := m.materializeColumn(ctx, rg, rg.ColumnChunks()[m.dataColToIndex[i]], rr)
+		if err != nil {
+			return r, err
+		}
+
+		for vi, value := range values {
+			chks, err := m.d.Decode(value.ByteArray(), mint, maxt)
+			if err != nil {
+				return r, errors.Wrap(err, "failed to decode chunks")
+			}
+			r[vi] = append(r[vi], chks...)
+		}
+	}
+
+	return r, nil
+}
+
+func (m *Materializer) materializeColumn(ctx context.Context, group parquet.RowGroup, cc parquet.ColumnChunk, rr []rowRange) ([]parquet.Value, error) {
+	if len(rr) == 0 {
+		return nil, nil
+	}
+
+	oidx, err := cc.OffsetIndex()
+	if err != nil {
+		return nil, errors.Wrap(err, "could not get offset index")
+	}
+
+	cidx, err := cc.ColumnIndex()
+	if err != nil {
+		return nil, errors.Wrap(err, "could not get column index")
+	}
+
+	pagesToRowsMap := make(map[int][]rowRange, len(rr))
+
+	for i := 0; i < cidx.NumPages(); i++ {
+		pageRowRange := rowRange{
+			from: oidx.FirstRowIndex(i),
+		}
+		pageRowRange.count = group.NumRows()
+
+		if i < oidx.NumPages()-1 {
+			pageRowRange.count = oidx.FirstRowIndex(i+1) - pageRowRange.from
+		}
+
+		for _, r := range rr {
+			if pageRowRange.Overlaps(r) {
+				pagesToRowsMap[i] = append(pagesToRowsMap[i], r)
+			}
+		}
+	}
+
+	r := make(map[rowRange][]parquet.Value, len(rr))
+	for _, v := range rr {
+		r[v] = []parquet.Value{}
+	}
+
+	errGroup := &errgroup.Group{}
+	errGroup.SetLimit(m.concurrency)
+
+	for _, p := range coalescePageRanges(pagesToRowsMap, oidx) {
+		errGroup.Go(func() error {
+			pgs := m.getPage(ctx, cc)
+			defer func() { _ = pgs.Close() }()
+			err := pgs.SeekToRow(p.rows[0].from)
+			if err != nil {
+				return errors.Wrap(err, "could not seek to row")
+			}
+
+			vi := new(valuesIterator)
+			remainingRr := p.rows
+			currentRr := remainingRr[0]
+			next := currentRr.from
+			remaining := currentRr.count
+			currentRow := currentRr.from
+
+			remainingRr = remainingRr[1:]
+			for len(remainingRr) > 0 || remaining > 0 {
+				page, err := pgs.ReadPage()
+				if err != nil {
+					return errors.Wrap(err, "could not read page")
+				}
+				vi.Reset(page)
+				for vi.Next() {
+					if currentRow == next {
+						r[currentRr] = append(r[currentRr], vi.At())
+						remaining--
+						if remaining > 0 {
+							next = next + 1
+						} else if len(remainingRr) > 0 {
+							currentRr = remainingRr[0]
+							next = currentRr.from
+							remaining = currentRr.count
+							remainingRr = remainingRr[1:]
+						}
+					}
+					currentRow++
+				}
+				parquet.Release(page)
+
+				if vi.Error() != nil {
+					return vi.Error()
+				}
+			}
+			return nil
+		})
+	}
+	err = errGroup.Wait()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to materialize columns")
+	}
+
+	values := make([]parquet.Value, 0, 1024)
```

Comment:
Why 1024? Sounds as if we could actually allocate the exact size. 

Very naive implementation:
```
	totalSize := 0
	for _, v := range rr {
		totalSize += int(v.count)
	}
	values := make([]parquet.Value, 0, totalSize)

```.

Reply:
done! :D